File: packages/serverless/lib/plugins/aws/bedrock-agentcore/examples/python/langgraph-multi-gateway/public_agent.py
"""
Public LangGraph agent with public gateway tools.

This agent uses the public gateway (NONE authorization) and has access
to the calculator tool which doesn't require authentication.

Uses plain streamablehttp_client since no IAM auth is required.
"""

import os
from typing import Annotated
from typing_extensions import TypedDict

from langchain.chat_models import init_chat_model
from langgraph.graph import StateGraph, START
from langgraph.graph.message import add_messages
from langgraph.prebuilt import ToolNode, tools_condition

from bedrock_agentcore.runtime import BedrockAgentCoreApp
from mcp import ClientSession
from mcp.client.streamable_http import streamablehttp_client
from langchain_mcp_adapters.tools import load_mcp_tools

app = BedrockAgentCoreApp()

llm = init_chat_model(
    "us.anthropic.claude-sonnet-4-5-20250929-v1:0",
    model_provider="bedrock_converse",
)

GATEWAY_URL = os.environ.get("BEDROCK_AGENTCORE_GATEWAY_URL")


# Define the conversation state
class State(TypedDict):
    messages: Annotated[list, add_messages]


async def chatbot_node(state: State, llm_with_tools):
    """Main chatbot node that invokes the LLM."""
    response = await llm_with_tools.ainvoke(state["messages"])
    return {"messages": [response]}


async def run_agent_with_gateway(user_message: str) -> str:
    """
    Run the LangGraph agent with tools discovered from public Gateway.
    Uses plain streamablehttp_client (no auth required for NONE authorizer).
    """
    if not GATEWAY_URL:
        print("No BEDROCK_AGENTCORE_GATEWAY_URL configured")
        graph_builder = StateGraph(State)
        graph_builder.add_node("chatbot", lambda state: {"messages": [llm.invoke(state["messages"])]})
        graph_builder.add_edge(START, "chatbot")
        graph = graph_builder.compile()

        result = graph.invoke({
            "messages": [{"role": "user", "content": user_message}]
        })
        return result['messages'][-1].content

    print(f"[Public Agent] Discovering tools from: {GATEWAY_URL}")

    # For NONE authorization, use plain streamablehttp_client (no auth)
    async with streamablehttp_client(url=GATEWAY_URL) as (read, write, session_id_callback):
        async with ClientSession(read, write) as session:
            await session.initialize()

            # Load MCP tools as LangChain tools
            tools = await load_mcp_tools(session)

            for tool in tools:
                print(f"  Discovered tool: {tool.name}")

            # Bind tools to LLM
            llm_with_tools = llm.bind_tools(tools)

            # Build the graph
            graph_builder = StateGraph(State)

            async def async_chatbot(state: State):
                return await chatbot_node(state, llm_with_tools)

            graph_builder.add_node("chatbot", async_chatbot)

            if tools:
                graph_builder.add_node("tools", ToolNode(tools=tools))
                graph_builder.add_conditional_edges("chatbot", tools_condition)
                graph_builder.add_edge("tools", "chatbot")

            graph_builder.add_edge(START, "chatbot")

            graph = graph_builder.compile()

            print(f"[Public Agent] Running with tools: {[t.name for t in tools]}")

            result = await graph.ainvoke({
                "messages": [{"role": "user", "content": user_message}]
            })

            return result['messages'][-1].content


@app.entrypoint
async def agent_invocation(payload, context):
    """
    Public agent entrypoint.
    Has access to calculator tool via public gateway (no auth required).
    """
    user_message = payload.get("prompt", "Hello! I'm the public agent with calculator access.")

    print(f"[Public Agent] Received: {user_message}")

    final_message = await run_agent_with_gateway(user_message)

    print(f"[Public Agent] Response: {final_message}")

    return {"result": final_message}


if __name__ == "__main__":
    app.run()

File: packages/serverless/lib/plugins/aws/bedrock-agentcore/examples/javascript/langgraph-multi-gateway/public-agent.js
/**
 * Public LangGraph agent with public gateway tools.
 *
 * This agent uses the public gateway (NONE authorization) and has access
 * to the calculator tool which doesn't require authentication.
 *
 * Uses plain StreamableHTTPClientTransport since no IAM auth is required.
 */

import { BedrockAgentCoreApp } from 'bedrock-agentcore/runtime'
import { ChatBedrockConverse } from '@langchain/aws'
import { StateGraph, START } from '@langchain/langgraph'
import { MessagesAnnotation } from '@langchain/langgraph'
import { ToolNode, toolsCondition } from '@langchain/langgraph/prebuilt'
import { Client } from '@modelcontextprotocol/sdk/client/index.js'
import { StreamableHTTPClientTransport } from '@modelcontextprotocol/sdk/client/streamableHttp.js'
import { loadMcpTools } from '@langchain/mcp-adapters'
import { z } from 'zod'

const GATEWAY_URL = process.env.BEDROCK_AGENTCORE_GATEWAY_URL
const AWS_REGION = process.env.AWS_REGION || 'us-east-1'

const llm = new ChatBedrockConverse({
  model: 'us.anthropic.claude-sonnet-4-5-20250929-v1:0',
  region: AWS_REGION,
})

/**
 * Connect to the public gateway (no auth required).
 */
async function createPublicMcpSession() {
  const transport = new StreamableHTTPClientTransport(new URL(GATEWAY_URL))

  const client = new Client({
    name: 'langgraph-public-agent',
    version: '1.0.0',
  })

  await client.connect(transport)
  return { client, transport }
}

/**
 * Run the agent with tools discovered from the public gateway.
 */
async function runAgentWithGateway(userMessage) {
  if (!GATEWAY_URL) {
    console.log('[Public Agent] No BEDROCK_AGENTCORE_GATEWAY_URL configured')
    const result = await llm.invoke([{ role: 'user', content: userMessage }])
    return result.content
  }

  console.log(`[Public Agent] Discovering tools from: ${GATEWAY_URL}`)

  const { client, transport } = await createPublicMcpSession()

  try {
    const tools = await loadMcpTools('public-gateway', client)

    for (const t of tools) {
      console.log(`  Discovered tool: ${t.name}`)
    }

    const llmWithTools = llm.bindTools(tools)

    const graphBuilder = new StateGraph(MessagesAnnotation)

    graphBuilder.addNode('chatbot', async (state) => {
      const response = await llmWithTools.invoke(state.messages)
      return { messages: [response] }
    })

    if (tools.length > 0) {
      graphBuilder.addNode('tools', new ToolNode(tools))
      graphBuilder.addConditionalEdges('chatbot', toolsCondition)
      graphBuilder.addEdge('tools', 'chatbot')
    }

    graphBuilder.addEdge(START, 'chatbot')
    const graph = graphBuilder.compile()

    console.log(
      `[Public Agent] Running with tools: ${tools.map((t) => t.name).join(', ')}`,
    )

    const result = await graph.invoke({
      messages: [{ role: 'user', content: userMessage }],
    })

    return result.messages[result.messages.length - 1].content
  } finally {
    await transport.close()
  }
}

const app = new BedrockAgentCoreApp({
  invocationHandler: {
    requestSchema: z.object({
      prompt: z.string().describe('The user message to process'),
    }),

    async process(request) {
      console.log(`[Public Agent] Received: ${request.prompt}`)

      const finalMessage = await runAgentWithGateway(request.prompt)

      console.log(`[Public Agent] Response: ${finalMessage}`)

      return JSON.stringify({ result: finalMessage })
    },
  },
})

app.run()
